### ISSUE TITLE
Why I moved to znapzend

### ISSUE THREAD

**Issue by redmop** [jimsalterjrs/sanoid#102]
@jimsalterjrs asked me about this, so I'm making a new issue. #8 

* Atomic snapshots - This is what had me start looking again. I use multiple virtual disks per VM. They are in the same tree (rpool/data/vms/snap-15/vm-100-disk-[1|2] for example). The new syncoid option takes care of part of it.
* Frequency - Missing the "frequent" and weekly timeframes.
* Pre/Post scripting - I didn't want to write a sanoid wrapper. With pre/post scripting, I can quince filesystems of VMs, rsync something, or pauses databases before the snapshots.
* Source handles expiring snapshots on the destination(s) - The source knows if the syncing is ok in znapzend, so it only expires snapshots on the source and destination only after all destinations are successfully synced. If Source syncs to DestA and DestB, and DestB is down, no snapshots are deleted on Source and DestB until B is caught up. This makes for happier off-sites through slow links.

Other nice features:
* Compression - Znapzend doesn't fix this either as it just uses ssh -c. Parallel compression for slow links is sexy. lzop is awesome for LAN. pigz for WAN links because more compression will speed the transfer rate, especially during the day when I limit bandwidth to 5 mbps. My sanoid repo has MANY compression types in it, but it's pretty ugly and I've not tried to do it a better way yet. Soon. I may just aim for adding pigz (which I will use) and then switches for custom ones.
* One off/triggered snapshots - zfsnap easily ties in with apt (and maybe rpm) and can do pre/post apt snapshots. This with etckeeper makes me very happy that I can recover from exploding updates. Also, the cron entry @reboot take-snapshot also makes me happy. It's more for a bookmark of when a reboot/crash happened, so I don't keep them long.
* Separating expiring and taking of snapshots - freeing up snapshot space takes IO. I want that after hours. sanoid can do it with some monkey business, znapzend can't do it at all. zfsnap does this very well.
* Sync separate from snapshot - I love that sanoid does this. Especially destinations. As in I can onsite sync every 15 min, and offsite right after work, then again after nightly processing.
* Pull and/or push syncing. Sanoid does this. znapzend does not. zfsnap doesn't sync at all.

I use syncoid for manual syncing pretty often, and findoid on my lxc containers from time to time to track changes. I do miss the monitoring.

I'm not really happy with znapzend (it's harder to teach/configure) but it hits all those need buttons but monitoring (I grep logs :/) and compression.

I use it on Debian (Proxmox) and Ubuntu (my workstation and laptop). I have snap every 15 min or 1 day, and sync as frequently (I used to do hourly, but meh, not much gain over 15m). I sync onsite and offsite. My offsite upstream is usually 10mbps, and I need to make sure I don't use it all, so I QoS it and limit it to 5mbps during peak times. Zfs is not the only backups I take. I also use backuppc, borg, vzdump (proxmox), and Crashplan (for now).

Wow, that's a lot. That kinda got away from me.

Sanoid is closer to how I would do it, coding wise. Znapzend is what I choose because I needed something now. Unfortunately, I just stuck with "good enough." I'm looking at a rather large project, so this will likely change soon.

**Comment by jimsalterjrs** [comment]
Most of this seems reasonable, but it's a big lump to digest. Would you be willing to open up separate issues for each feature request?

**Comment by ewwhite** [comment]
I don't think you should try to shape your product to fit such narrow use cases.

Edmund White

On Jul 11, 2017, at 6:58 PM, Jim Salter <notifications@github.com<mailto:notifications@github.com>> wrote:


Most of this seems reasonable, but it's a big lump to digest. Would you be willing to open up separate issues for each feature request?

**Comment by redmop** [comment]
Sure. I'll take care of that tomorrow morning. It's punch out time.

**Comment by jimsalterjrs** [comment]
I don't know that *all* of it is a yes, but it's easier to deal with on a case by case. There's some merit to making atomic snapshots across datasets possible (where they ARE possible, since this is interesting like inherently limited to child datasets and no exclusions), and extra compression algorithms is no biggie. I had already been planning to implement pigz in particular.

**Comment by redmop** [comment]
* Frequency is covered here: #75 
* Separating expiring and taking of snapshots - Already done with --take-snapshots and --prune-snapshots if you make separate cron jobs for them.
* Push/Pull syncing is already done.

**Comment by redmop** [comment]
As I have broken this out, I suppose we can close this one.

**Comment by mailinglists35** [comment]
I don't have so many cores to test, but in my 2 cores x 2 threads cpu lz4 beats the sh*t out of pigz (and uses much less cpu while pigz uses 100% allo four logical cpus)

lz4: 388MiB/s, one cpu 100%

pigz: 74.1MiB/s four logical cpus 100%

**Comment by ewwhite** [comment]
Correct,

Pigz doesn’t make sense here.

**Comment by redmop** [comment]
pigz would be used for WAN links where the line speed is the bottleneck. Better compression ratios are more important when your line speed is 10 MiBs as it helps push more data through the link. for LAN, lzop/lz4 is better.

**Comment by jimsalterjrs** [comment]
Yeah, pigz should not be the *default* but it's a sensible option in a lot of settings I personally deal with. I have a lot of sites with 5 Mbps upload; chewing through eight+ threads in return for another 15-20% compression is frequently a win there.

It's also trivial to implement; we've already got pluggable compression to switch between lzop and gzip.

**Comment by jimsalterjrs** [comment]
On further review, pigz is pretty pointless on a slow internet link; simple gzip produces more than enough bandwidth for a 100 Mbps link on the hardware specs I typically use.  However, it actually _might_ make sense over gigabit LAN links if you're in a hurry, weirdly enough.

I tested using the first 2GB of data from the boot drive on my workstation:

~~~~
root@banshee:~/git/sanoid# gzip -3 < /banshee/test/test.bin | pv > /dev/null
 516MiB 0:00:28 [18.1MiB/s] [                           <=>                    ]
root@banshee:~/git/sanoid# pigz -3 < /banshee/test/test.bin | pv > /dev/null
 514MiB 0:00:06 [75.2MiB/s] [      <=>                                         ]
root@banshee:~/git/sanoid# pigz -9 < /banshee/test/test.bin | pv > /dev/null
 494MiB 0:00:27 [17.8MiB/s] [                          <=>                     ]
root@banshee:~/git/sanoid# lzop < /banshee/test/test.bin | pv > /dev/null
 645MiB 0:00:04 [ 154MiB/s] [    <=>                                           ]
~~~~

`gzip -3` already produces more than enough compressed throughput to saturate a 100 Mbps WAN link - and then some.  `pigz -9` is about the same compressed throughput, but around 5% more efficient.  There's some question as to whether it's _worth_ cutting 5% off your transmission time in return for `pigz` chewing through all available threads, of course - but there's very little cost (five lines of code) in making the option available.

`pigz -3` looks like an interesting option for gigabit LAN links; it produces enough compressed throughput to saturate or nearly saturate most real-world fast LANs, with a whopping 20% improvement in compressed data size over `lzop`.  I wouldn't want to make it the _default_, for several reasons, but I think it looks easily interesting enough to make available, again with a total cost of four lines of code.

Also keep in mind that while `pigz` will chew through all available threads at 100% utilization _if unthrottled_, in many of the cases with WAN links it will end up being more like 10%-50% utilization across all threads.  Whether this is preferable or not is debatable; it's maybe better thermal spreading across cores but might impact individual VMs noticeably where `gzip` wouldn't.  It's all kinda debatable honestly; I'm still thinking "not a good _default_, but worth it as a pluggable option."

**Comment by jimsalterjrs** [comment]
I've pushed these options as `compress=pigz-fast` and `compress=pigz-slow` in 1.4.10 in master.

Default compression remains `lzop` for remote replication, with forced `none` on localhost-only replication.

**Comment by redmop** [comment]
Actually, using all the cores will use more power as they won't be able to drop to low power.
